app/services/ReservationService.java

```diff
@@ -0,0 +1,191 @@
+            boolean bool = false;
+        for(Reservation reservation : reservations) {
+            if ((!Objects.equals(reservation.getReservationsStart(), selectedTimeAndDate)) && (reservation.getReservationsStart() != selectedTimeAndDate + HALF_HOUR)
+                    && (reservation.getReservationsStart() != selectedTimeAndDate + HOUR) && (reservation.getReservationsStart() != selectedTimeAndDate + (HOUR + HALF_HOUR))
+                    && (reservation.getReservationsStart() != selectedTimeAndDate - HALF_HOUR)
+                    && (reservation.getReservationsStart() != selectedTimeAndDate - HOUR) && (reservation.getReservationsStart() != selectedTimeAndDate - (HOUR + HALF_HOUR))) {
+                bool = true;
+            }
+        }
+        return bool;
+
+    }
+
+
+    public List getBestTimes(String date, Long dateAndTime, Long restaurantId, int people) {
+
+        Timestamp t1 = Timestamp.valueOf(date + " 08:00:00.0");
+        Timestamp t2 = Timestamp.valueOf(date + " 21:00:00.0");
+        FIRST_RESERVATION = t1.getTime();
+        LAST_RESERVATION = t2.getTime();
+        ArrayList<String> bestTime = new ArrayList<>();
+        List<RestaurantTable> tables = tableRepository.getTablesForSelectedNumberOfPeople(restaurantId, people);
+        List<Reservation> reservations = reservationRepository.getAllReservationsOnSelectedDate(date, tables);
+
+
+        if(reservations.size() == 0){
+            return getFirstReservation(dateAndTime);
+        } else {
+            for (RestaurantTable table: tables) {
+                if(!reservations.contains(table)){
+                    return getFirstReservation(dateAndTime);
+                }
+            }
+        }
+
+        Long timeToCheckUp = dateAndTime;
+        Long timeToCheckDown = dateAndTime;
+
+        int loopBreaker = 0;
+
+        while (bestTime.size() < 4) {
+            loopBreaker++;
+
+            if (timeToCheckUp != null) {
+                if (hasFreeTables(reservations, timeToCheckUp)) {
+                    Date date2 = new Date(timeToCheckUp);
+                    DateFormat formatter = new SimpleDateFormat("HH:mm");
+                    bestTime.add(formatter.format(date2));
+
+                }
+            }
+
+            timeToCheckUp = stepUp(timeToCheckUp);
+            timeToCheckDown = stepDown(timeToCheckDown);
+
+            if (timeToCheckDown != null) {
+                if (hasFreeTables(reservations, timeToCheckDown)) {
+                    Date date2 = new Date(timeToCheckDown);
+                    DateFormat formatter = new SimpleDateFormat("HH:mm");
+                    bestTime.add(formatter.format(date2));
+
+                }
+            }
+
+            if(loopBreaker > 30){
+                break;
+
+            }
+        }
+
+        return bestTime;
+
+    }
+
+
+    private Long stepUp(Long time) {
+        if (time != null) {
+            if (time <= LAST_RESERVATION) {
```

Comment:
`LAST_RESERVATION ` is neither final, nor static. There should be another argument in this method `maxTime` that should be passed to this method (same below for `stepDown` -> `minTime`).